### ISSUE TITLE
security: Password type stored as clear text

### ISSUE THREAD

**Issue by Scapal** [dgraph-io/dgraph#2623]
Even if the `Password` type is secured by being unreadable, only verifiable, it is stored as plain text.

By just altering the schema an changing the type back to `string`, you can dump all the passwords. This is very **wrong**.

You should [scrypt](https://godoc.org/golang.org/x/crypto/scrypt) the password using the node `uid` and the predicate name as source for the salt.

This can be done without changing the usage behavior.

**Comment by MichelDiz** [comment]
Password type is encrypted. You have the ability to transform any Type to other types so you free turn Password type to plain text tho.

Check: https://github.com/dgraph-io/dgraph/blob/06ea4c545bc3cf0ed730327557dbff96406e75f8/types/password.go#L25

If you need to have full control over Password, please use other approaches and write with String type. This feature is just to make the DB user's life easier. If the user need something more complex better by DIY. More safe and controleble by their end.

@manishrjain Quick question, though it is encrypted. Is there a need to do something for this case? Otherwise it can be closed.

**Comment by srfrog** [comment]
Perhaps not let passwords to be altered to other types ? It would be a new feature, and a good one IMO.

**Comment by MichelDiz** [comment]
@srfrog Gus, I think we should have a Lock to alter. Something like a PIN or a password. That might help. But also prevent from using Convert func from PasswordID to string. For in theory not even the DB maintainer should have access to the plain text. For that would be an invasion of privacy.

What do you think Gus?

@Scapal It was encrypted for sure, the problem is that you're free to convert tho. Dgraph is able to freely convert a Password type, since it is the one that generates the same ones. Soon it is not a Bug, as we can see clearly in the code the open possibility. So it's not a glitch. Whoever wrote the code knew what he was doing.

So that's why it's a feature.

Cheers.

**Comment by srfrog** [comment]
i agree, but we need to support the existing deployment as they are. i think for now the default would have to stay plain text, and introduce hashing function(s) as options. ie.,
```
secret: password @hash(scrypt) .
```
also, what would happen when you change from say plain to script? i think we have a good idea here, we need to develop it a bit more. what do you think?

**Comment by manishrjain** [comment]
Hey guys, I want to weigh in here:

The existing functionality is right (looks buggy) but needs tweaks.

What's right: We never store the password in plain text (or at least that's the logic, disregarding any bugs for now). We're not going to change that behavior. A password type should not be 

What we should tweak: We should not allow a password type to be altered, i.e. a default type can't be converted to `passwd` type, and vice-versa. It can only be "created" and "dropped", i.e. either a predicate should be `passwd` type to begin with (before the first mutation), or it should be deleted. It should not be converted. Having said that, this bug needs an investigation to figure out how is it possible for us to store data encrypted, and it getting returned as plaintext.

**Comment by Scapal** [comment]
When creating a `secret` predicate of type `Password` and perform an export, here is what I get:

**rdf file**:
```
<_:uid7> <name> "Alice"^^<xs:string> .
<_:uid7> <secret> "Password1234"^^<xs:string> .
```

**schema file**:
```
secret:password . 
```

It seems the `secret` is never going through type conversion.

**Comment by Scapal** [comment]
ok, the `<xs:string>` is legit in the export but it should be the encrypted Base64 value. The `Convert()`from `password` to `string` just copies the string.
https://github.com/dgraph-io/dgraph/blob/06ea4c545bc3cf0ed730327557dbff96406e75f8/types/conversion.go#L312

**Comment by Scapal** [comment]
The problem only occurs with json mutations.

```
{
  set {
    <0x7> <secret> "Password1234" .
  }
}
```
properly encodes the secret:

```
<_:uid7> <secret> "$2a$10$FWbP4pN0jcr5mRU7iLtEAexR/heMaKRNARGoswOW2njHKYPSKSkT."^^<xs:string> .
```

**Comment by srfrog** [comment]
@Scapal I have a PR to fix this. see if it resolves your issue #2627
